File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/MarketElementReason.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/MarketElementReason.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class MarketElementReason extends Element
{
    private final MarketElementString _source, _category, _description;
    private final MarketElementInt _errorCode;

    public enum ReasonTypeEnum { badSecurity, badField }

    public MarketElementReason(ReasonTypeEnum reasonType)
    {
        if (reasonType == ReasonTypeEnum.badSecurity)
        {
            this._source = new MarketElementString("source", String.format("%s%s@%s", RandomDataGenerator.randomString(5).toLowerCase(), RandomDataGenerator.randomInt(9), RandomDataGenerator.randomInt(99)));
            this._errorCode = new MarketElementInt("errorCode", RandomDataGenerator.randomInt(99));
            this._category = new MarketElementString("category", "BAD_SEC");
            this._description = new MarketElementString("description", "Invalid security, rcode = -1");
        }
        else //if (reasonType == ReasonTypeEnum.badField)
        {
            this._source = new MarketElementString("source", String.format("%s@%s", RandomDataGenerator.randomString(7).toLowerCase(), RandomDataGenerator.randomInt(999)));
            this._errorCode = new MarketElementInt("errorCode", RandomDataGenerator.randomInt(99));
            this._category = new MarketElementString("category", "BAD_FLD");
            this._description = new MarketElementString("description", "Unknown Field");
        }
    }
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("ErrorInfo"));
    }
	
    public Name name() throws Exception
    {
    	return new Name("reason");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 4;
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    public int getElementAsInt32(String name) throws Exception
    {
    	return this.getElement(name).getValueAsInt32();
    }
    
    public String getElementAsString(String name) throws Exception
    {
    	return this.getElement(name).getValueAsString();
    }
        public Element getElement(String name) throws Exception
    {
    	if(name.equals("source"))
    		return this._source;
    	
    	else if(name.equals(this._errorCode.name().toString()))
    		return this._errorCode;
    	
    	else if(name.equals("category"))
    		return this._category;
    	
    	else if(name.equals("description"))
    		return this._description;
    	
    	else
    		throw new Exception("name not recognized. names are case-sensitive.");
    }
    
    public boolean hasElement(String name, boolean excludeNullElements) throws Exception
    {
    	return this.hasElement(name);
    }
    
    public boolean hasElement(String name) throws Exception
    {
    	return 
    			name.equals("source") || 
    			name.equals(this._errorCode.name().toString()) ||
    			name.equals("category") ||
    			name.equals("description");
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();
        
        result.append(String.format("%s%s = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));
        result.append(this._source.prettyPrint(tabIndent + 1));
        result.append(this._errorCode.prettyPrint(tabIndent + 1));
        result.append(this._category.prettyPrint(tabIndent + 1));
        result.append(this._description.prettyPrint(tabIndent + 1));
        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));

        return result;
    }

}

File: BEmu/MarketDataRequest/MarketElementReason.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="MarketDataRequest/MarketElementReason.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.MarketDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class MarketElementReason : Element
    {
        private readonly MarketElementString _source, _category, _description;
        private readonly MarketElementInt _errorCode;

        public enum ReasonTypeEnum { badSecurity, badField }

        public MarketElementReason(ReasonTypeEnum reasonType)
        {
            if (reasonType == ReasonTypeEnum.badSecurity)
            {
                this._source = new MarketElementString("source", string.Format("{0}{1}@{2}", Types.RandomDataGenerator.RandomString(5).ToLower(), Types.RandomDataGenerator.RandomInt(9), Types.RandomDataGenerator.RandomInt(99)));
                this._errorCode = new MarketElementInt("errorCode", Types.RandomDataGenerator.RandomInt(99));
                this._category = new MarketElementString("category", "BAD_SEC");
                this._description = new MarketElementString("description", "Invalid security, rcode = -1");
            }
            else if (reasonType == ReasonTypeEnum.badField)
            {
                this._source = new MarketElementString("source", string.Format("{0}@{1}", Types.RandomDataGenerator.RandomString(7).ToLower(), Types.RandomDataGenerator.RandomInt(999)));
                this._errorCode = new MarketElementInt("errorCode", Types.RandomDataGenerator.RandomInt(99));
                this._category = new MarketElementString("category", "BAD_FLD");
                this._description = new MarketElementString("description", "Unknown Field");
            }
        }

        public override SchemaTypeDefinition TypeDefinition { get { return new SchemaTypeDefinition(this.Datatype, new Name("ErrorInfo")); } }
        public override int NumElements { get { return 4; } }
        public override int NumValues { get { return 1; } }
        public override Name Name { get { return new Name("reason"); } }
        public override bool IsArray { get { return false; } }
        public override bool IsComplexType { get { return true; } }
        public override bool IsNull { get { return false; } }

        public override IEnumerable<Element> Elements
        {
            get
            {
                yield return this._source;
                yield return this._errorCode;
                yield return this._category;
                yield return this._description;
            }
        }

        public override string GetElementAsString(string name)
        {
            return this[name].GetValueAsString();
        }

        public override int GetElementAsInt32(string name)
        {
            return this[name].GetValueAsInt32();
        }

        public override Element GetElement(string name)
        {
            return this[name];
        }

        public override object GetValue()
        {
            return null;
        }

        public override object GetValue(int index)
        {
            return null;
        }

        public override object this[int index]
        {
            get
            {
                return null;
            }
        }

        public override Element this[string name]
        {
            get
            {
                string strName = name;
                foreach (var item in this.Elements)
                {
                    if (item.Name.ToString() == strName)
                        return item;
                }
                return base[name];
            }
        }

        public override object this[string name, int index]
        {
            get
            {
                if (index == 0)
                    return this[name];
                else
                    throw new IndexOutOfRangeException();
            }
        }

        public override bool HasElement(string name, bool excludeNullElements = false)
        {
            foreach (var item in this.Elements)
            {
                if (item.Name.ToString() == name)
                    return true;
            }
            return false;
        }

        internal override StringBuilder PrettyPrint(int tabIndent, bool surroundValueWithQuotes = false)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}reason = {{{1}", tabs, Environment.NewLine);
            foreach (var item in this.Elements)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);
            return result;
        }

    }
}
